packages/flutter/lib/src/widgets/app_model.dart

```diff
@@ -0,0 +1,203 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'framework.dart';
+import 'inherited_model.dart';
+
+/// The type of the [AppModel.getValue] `init` parameter.
+///
+/// This callback is used to lazily create the initial value for
+/// an [AppModel] keyword.
+typedef AppModelInitCallback<T> = T Function();
+
+/// Enables sharing key/value data with its `child` and all of the
+/// child's descendants.
+///
+/// - `AppModel.getValue(context, key, initCallback)` creates a dependency
+/// on the key and returns the value for the key from the shared data table.
+/// If no value exists for key then the initCallback is used to create
+/// the initial value.
+///
+/// - `AppModel.setValue(context, key, value)` changes the value of an entry
+/// in the shared data table and forces widgets that depend on that entry
+/// to be rebuilt.
+///
+/// A widget whose build method uses AppModel.getValue(context,
+/// keyword, initCallback) creates a dependency on the AppModel. When
+/// the value of keyword changes with AppModel.setValue(), the widget
+/// will be rebuilt. The values managed by the AppModel are expected
+/// to be immutable: intrinsic changes to values will not cause
+/// dependent widgets to be rebuilt.
+///
+/// An instance of this widget is created automatically by [WidgetsApp].
+///
+/// There are many ways to share data with a widget subtree. This
+/// class is based on [InheritedModel], which is an [InheritedWidget].
+/// It's intended to be used by packages that need to share a modest
+/// number of values among their own components. It obviates the
+/// need for package users to instantiate package-specific
+/// data sharing "umbrella" widgets.
+///
+/// AppModel is not intended to be a substitute for Provider or any of
+/// the other general purpose application state systems. AppModel is
+/// for situations where a package's custom widgets need to share one
+/// or a handful of immutable data objects that can be lazily
+/// initialized. It exists so that packages like that can deliver
+/// custom widgets without requiring the developer to add a
+/// package-specific umbrella widget to their application.
```

Comment:
nit: the sentence about the umbrella widget appears to be a repetition of the last sentence in the previous paragraph.